### PR TITLE
[TAS-794] ✨ Integrate epub.js into reader page

### PR DESCRIPTION
**Pull request by WeiJun0827**
https://github.com/likecoin/liker-land/assets/33746295/ff206c6f-8a84-4377-9fa9-2f3ed06f8835

https://github.com/likecoin/liker-land/assets/33746295/31b3878f-bc94-4b07-a8f3-6c1ca60bdfdf

### CODE REVIEW

**Comment by WeiJun0827** [comment]
QA link:
https://rinkeby.liker.land/zh-Hant/reader?download=1&classId=likenft13vwv5juw9cjktu8r6wqerlmcsqw3d2fyd5ckghmml7gknmmd69aswjg6t0&format=epub&src=https%3A%2F%2Fipfs.io%2Fipfs%2FQmXz6rCL7hvbNVMcvrn8HzSRSquSRheUAQUGXCRL1A2Fjz%3Fname%3Danimal_farm.epub

File: src/pages/reader.vue [hunk 1]
```diff
@@ -116,9 +163,68 @@ export default {
...
+  height: 90%;
+  z-index: 1;
+  left: 50%;
+  margin-left: -1px;
+  top: 5%;
+  opacity: 0.15;
+  box-shadow: -2px 0 15px rgba(0, 0, 0, 1);
+  content: '';
+}
+</style>
```
Comment (nwingt) [inline src/pages/reader.vue:230]: not sure if it is better to have separate pages for PDF and Epub readers before the logic and styling are intertwined tgt

e.g.
```js
// Before
'/reader'

// After
'/reader/pdf'
'/reader/epub'
```

File: src/pages/reader/epub.vue [hunk 2]
```diff
@@ -94,7 +86,7 @@ export default {
       return this.$route.query.classId;
     },
     format() {
-      return this.$route.query.format || 'pdf';
+      return this.$route.query.format || 'epub';
     },
```
Comment (nwingt) [inline src/pages/reader/epub.vue]: No need after splitting the page?

File: src/pages/reader/pdf.vue [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    bodyAttrs: { class: 'overflow-hidden' },
+  },
+  computed: {
+    ...mapGetters(['getHomeRoute']),
+    classId() {
+      return this.$route.query.classId;
+    },
+    format() {
+      return this.$route.query.format || 'pdf';
+    },
```
Comment (nwingt) [inline src/pages/reader/pdf.vue]: No need after splitting the page?

**Review by nwingt — COMMENTED** [review]
Can put common logic to `pages/reader/reader.vue` wtih `<NuxtChild v-if="CONDITION" />` where `CONDITION` is set to true after rendering is ready, e.g. login check

File: src/components/NFTViewOptionList.vue [hunk 4]
```diff
@@ -183,50 +184,41 @@ export default {
...
-      return `${this.getFilenameFromURL(url) || 'content'}`;
-    },
+    getFilenameFromURL,
+    getDownloadFilenameFromURL,
     handleClickViewContent() {
       this.$emit('view-content');
     },
+    async passCORSPolicy(url) {
+      try {
+        await this.$axios.head(url);
```
Comment (williamchong) [inline src/components/NFTViewOptionList.vue:194]: Not all GET resources support HEAD though, safest bet is try a GET and ignore the body after we get the 200 status

from chatGPT

```
axios.get('https://example.com/api/endpoint', { responseType: 'stream' })
  .then(response => {
    if (response.status === 200) {
      console.log('GET request succeeded');
      // You can choose to abort the request here if you don't need the response body
      response.data.destroy();
    } else {
      console.log('GET request failed');
    }
  })
  .catch(error => {
    console.log('An error occurred:', error);
  });
  ```

Comment (WeiJun0827) [reply src/components/NFTViewOptionList.vue]: Nice trick, thanks!!

File: src/components/NFTViewOptionList.vue [hunk 5]
```diff
@@ -183,50 +184,44 @@ export default {
...
+    getDownloadFilenameFromURL,
     handleClickViewContent() {
       this.$emit('view-content');
     },
+    async passCORSPolicy(url) {
+      try {
+        const res = await this.$axios.get(url, { responseType: 'stream' });
+        if (res.status === 200) {
+          res.data.destroy();
+        }
```
Comment (williamchong) [inline src/components/NFTViewOptionList.vue:197]: should accept 2xx and 3xx

File: src/components/NFTViewOptionList.vue [hunk 6]
```diff
@@ -183,50 +184,44 @@ export default {
...
+                download: this.isContentDownloadable ? '1' : '0',
+                classId: this.classId,
+                format: type,
+                src: url,
+              },
+            })
+          );
+        } else {
+          // bypass CORS
+          window.location.href = url;
```
Comment (williamchong) [inline src/components/NFTViewOptionList.vue:224]: Should not allow download if `isContentDownloadable` is false, just throw error

**Review by williamchong — CHANGES_REQUESTED** [review]

File: src/components/NFTViewOptionList.vue [hunk 7]
```diff
@@ -183,50 +184,44 @@ export default {
...
-          if (error.message === 'Network Error') {
-            // bypass CORS
-            window.location.href = url;
-          } else {
-            // eslint-disable-next-line no-console
-            console.error(error);
-            this.alertPromptError(this.$t('nft_download_content_error'));
-          }
+        const passCORS = await this.passCORSPolicy(url);
+        if (passCORS) {
```
Comment (williamchong) [inline src/components/NFTViewOptionList.vue:210]: On second thought, why check CORS here? We can always handle it in reader page for pdf and epub using the cors proxy cloud function

**Review by williamchong — CHANGES_REQUESTED** [review]

**Review by williamchong — APPROVED** [review]

**Review by nwingt — APPROVED** [review]

**Pull request merged**